### PR TITLE
fix: template literal dynamic import

### PR DESCRIPTION
**Pull request by faga295**
### <samp>🤖 Generated by Copilot at 7b5bf06</samp>

This pull request adds support for dynamic imports with template literals in JavaScript modules. It modifies the `rspack_plugin_javascript` and `rspack_core` crates to handle different cases of dynamic import arguments. It also adds test fixtures and snapshots for the dynamic import feature in the `rspack` and `rspack/tests` packages. Additionally, it applies some code style preferences to the existing test cases.

## Related issue (if exists)
#2687 

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at 7b5bf06</samp>

*  Support dynamic imports with template literals as arguments in rspack ([link](https://github.com/web-infra-dev/rspack/pull/2732/files?diff=unified&w=0#diff-65726b45e3f83132a525b4fa2c0f7dc69dc45d35e167175a1e8912bda28d82cfL143-R143), [link](https://github.com/web-infra-dev/rspack/pull/2732/files?diff=unified&w=0#diff-8281ab635aae318f63d8e3d743577a933a15bda2136698d2900e8bd8c711d2faL14-R14), [link](https://github.com/web-infra-dev/rspack/pull/2732/files?diff=unified&w=0#diff-8281ab635aae318f63d8e3d743577a933a15bda2136698d2900e8bd8c711d2faL126-R173))
* Add test fixtures for the dynamic import feature of rspack ([link](https://github.com/web-infra-dev/rspack/pull/2732/files?diff=unified&w=0#diff-3a6288d6314d317a4d96cddcf94d3fc655ce4bbade63cc83e05d5e4249025226L1-R34), [link](https://github.com/web-infra-dev/rspack/pull/2732/files?diff=unified&w=0#diff-d3490d5baea14378aeb0fede652739dbc6cc4d94703e5a6a5088b83bd7d63707L1-R3), [link](https://github.com/web-infra-dev/rspack/pull/2732/files?diff=unified&w=0#diff-8642fe546756652993bbad0a22e7c42fb54a1154e0868444af337d67f6cad31eL1-R8), [link](https://github.com/web-infra-dev/rspack/pull/2732/files?diff=unified&w=0#diff-d597a8272e28d97a660b3411cd057661c802073139549a58baf266489e3ea9d0L1-L-1), [link](https://github.com/web-infra-dev/rspack/pull/2732/files?diff=unified&w=0#diff-1d29f30a0dd16581f3391d5137b376dd58cb5b3725547625347c3e2494e64016L1-L-1), [link](https://github.com/web-infra-dev/rspack/pull/2732/files?diff=unified&w=0#diff-b9203325b31b1430b10275b7f30038beeb6c9f129d065cd2bdd5d92ce3a31a39L1-L-1))
* Apply code style preference for single quotes and parentheses in test files ([link](https://github.com/web-infra-dev/rspack/pull/2732/files?diff=unified&w=0#diff-f71cb053440183ee326badf3ab696524cc93c1f4d8f03589c88ba9abb47882faL1-R1), [link](https://github.com/web-infra-dev/rspack/pull/2732/files?diff=unified&w=0#diff-5d4cc5185aaf60d0fe4229e949981cc1f2b00f95710deeec24bf738f659292bfL7-R12), [link](https://github.com/web-infra-dev/rspack/pull/2732/files?diff=unified&w=0#diff-3ccd42e1022899055a87b963d294f1fccd3e12b81a081aebdced28d7490c9626L1-R1), [link](https://github.com/web-infra-dev/rspack/pull/2732/files?diff=unified&w=0#diff-e5375a42b2b2dea0801ce6d267ed890ecbc312fc8b92914e4b5a05bc69d3ef7dL7-R12), [link](https://github.com/web-infra-dev/rspack/pull/2732/files?diff=unified&w=0#diff-f69e0c13c8cf2d0c65cec548efcdc7e08fe091acf818a687c613ec2f19a85f07L55-R55))

</details>

### CODE REVIEW

File: crates/rspack_plugin_javascript/Cargo.toml [hunk 1]
```diff
@@ -48,5 +48,5 @@ swc_emotion = { workspace = true }
 swc_plugin_import = { workspace = true }
 tokio = { workspace = true, features = ["rt"] }
 tracing = { workspace = true }
+url = "2.3.1"
```
[the remark this answers is not in the document]
Comment (faga295) [reply crates/rspack_plugin_javascript/Cargo.toml:51]: sorry, it's not expected

File: crates/rspack_plugin_javascript/src/visitors/dependency/scanner.rs [hunk 2]
```diff
@@ -123,30 +123,55 @@ impl DependencyScanner<'_> {
...
+                    .to_string(),
+                );
+                self.add_dependency(box EsmDynamicImportDependency::new(
+                  request,
+                  Some(node.span.into()),
+                  as_parent_path(ast_path),
+                  chunk_name,
+                ));
+              }
+              _ => {
```
[the remark this answers is not in the document]
Comment (faga295) [reply crates/rspack_plugin_javascript/src/visitors/dependency/scanner.rs:155]: `_` in there represents `length > 1`

File: crates/rspack_plugin_javascript/src/visitors/dependency/scanner.rs [hunk 3]
```diff
@@ -123,30 +123,55 @@ impl DependencyScanner<'_> {
...
+            Expr::Lit(Lit::Str(imported)) => {
+              let chunk_name = self.try_extract_webpack_chunk_name(&imported.span);
+              self.add_dependency(box EsmDynamicImportDependency::new(
+                imported.value.clone(),
+                Some(node.span.into()),
+                as_parent_path(ast_path),
+                chunk_name,
+              ));
+            }
+            Expr::Tpl(tpl) => match tpl.quasis.len() {
```
[the remark this answers is not in the document]
Comment (faga295) [reply crates/rspack_plugin_javascript/src/visitors/dependency/scanner.rs:136]: Should I ignore that warning?

[the remark this answers is not in the document]
Comment (hyf0) [reply crates/rspack_plugin_javascript/src/visitors/dependency/scanner.rs:136]: Could you post the warning here? I'm curious.

[the remark this answers is not in the document]
Comment (hyf0) [reply crates/rspack_plugin_javascript/src/visitors/dependency/scanner.rs:136]: Get it. I'm suggesting the below changes.

Rust supports match pattern with guards
```rust
let t = Some([1, 2])
match t {
  Some(v) if v.len() == 1 => {
  },
  Some(v) if v.len() == 2 => {
  },
  _ => {},
}
```

File: crates/rspack_plugin_javascript/src/visitors/dependency/scanner.rs [hunk 4]
```diff
@@ -123,30 +123,55 @@ impl DependencyScanner<'_> {
...
+            Expr::Lit(Lit::Str(imported)) => {
+              let chunk_name = self.try_extract_webpack_chunk_name(&imported.span);
+              self.add_dependency(box EsmDynamicImportDependency::new(
+                imported.value.clone(),
+                Some(node.span.into()),
+                as_parent_path(ast_path),
+                chunk_name,
+              ));
+            }
+            Expr::Tpl(tpl) => match Some(&tpl.quasis) {
```
Comment (hyf0) [inline crates/rspack_plugin_javascript/src/visitors/dependency/scanner.rs:136]: I might cause some misunderstanding to you :)

I'm suggesting changes like this
```rust
match dyn_imported.expr.as_ref() {
   Expr::Tpl( { quasis }) if quasis.len() == 1 => {
     // ..
   },
   Expr::Tpl(Tpl{ quasis, .. }) if quasis.len() > 1 => {
     // ..
   },
}
```

- You don't need `Option` to use pattern matching with guard
- `Tpl` in `Tpl{ quasis, .. }` is imported from `swc_core::ecma::ast::Tpl`

**Review by hyf0 — APPROVED** [review]
LGTM. Let's wait for the review of @underfin.

**Review by underfin — APPROVED** [review]

**Pull request merged**